### PR TITLE
[AMS-452][AMS-527] new element type vectorImage

### PR DESCRIPTION
**Pull request by xakep139**
`image` element type was renamed to `bitmapImage`

### CODE REVIEW

File: src/VStore/Sessions/SessionManagementService.cs [hunk 1]
```diff
@@ -128,17 +131,17 @@ public sealed class SessionManagementService
...
+                    Key = key,
+                    ContentType = contentType
+                };
             var metadataWrapper = MetadataCollectionWrapper.For(request.Metadata);
             metadataWrapper.Write(MetadataElement.Filename, fileName);
 
             var uploadResponse = await _amazonS3.InitiateMultipartUploadAsync(request);
 
-            return new MultipartUploadSession(sessionId, fileKey, uploadResponse.UploadId);
+            return new MultipartUploadSession(sessionId, sessionDescriptor, fileKey, uploadResponse.UploadId);
```
Comment (xakep139) [inline src/VStore/Sessions/SessionManagementService.cs]: Save session descriptor for optimization

File: src/VStore/Sessions/SessionManagementService.cs [hunk 2]
```diff
@@ -192,7 +195,7 @@ public sealed class SessionManagementService
                                          });
             uploadSession.Complete();
 
-            SessionDescriptor sessionDescriptor = await GetSessionDescriptor(uploadSession.SessionId);
```
Comment (xakep139) [inline src/VStore/Sessions/SessionManagementService.cs]: Don't need to load again

File: src/VStore.Host/Controllers/SessionsController.cs [hunk 3]
```diff
@@ -169,6 +169,7 @@ public SessionsController(SessionManagementService sessionManagementService, ILo
         [ProducesResponseType(typeof(UploadedFileValue), 201)]
         [ProducesResponseType(typeof(string), 400)]
         [ProducesResponseType(typeof(string), 404)]
+        [ProducesResponseType(typeof(string), 410)]
```
Comment (xakep139) [inline src/VStore.Host/Controllers/SessionsController.cs]: Gone result if session has expired

**Comment by xakep139** [comment]
@denisivan0v, could we save file initially with extension instead of copying it in `CompleteMultipartUpload()`?
 https://github.com/2gis/nuclear-vstore/blob/master/src/VStore/Sessions/SessionManagementService.cs#L215

File: src/VStore/Sessions/SessionManagementService.cs (hunk 1 above)
Comment (xakep139) [inline src/VStore/Sessions/SessionManagementService.cs]: Additionally, we could load template element descriptor by calling `GetElementDescriptor()` only once. At the moment it loads twice

**Comment by xakep139** [comment]
> @denisivan0v, could we save file initially with extension instead of copying it in CompleteMultipartUpload()?
https://github.com/2gis/nuclear-vstore/blob/master/src/VStore/Sessions/SessionManagementService.cs#L215

Discussed

File: src/VStore/Sessions/SessionManagementService.cs (hunk 2 above)
Comment (denisivan0v) [inline src/VStore/Sessions/SessionManagementService.cs]: See the comment above

File: src/VStore/Sessions/SessionManagementService.cs [hunk 4]
```diff
@@ -138,35 +145,30 @@ public sealed class SessionManagementService
...
+        public async Task UploadFilePart(MultipartUploadSession uploadSession, Stream inputStream, int templateCode)
         {
-            using (var memory = new MemoryStream())
-            {
-                await inputStream.CopyToAsync(memory);
-                memory.Position = 0;
-
-                if (uploadSession.NextPartNumber == 1)
-                {
-                    SessionDescriptor sessionDescriptor = await GetSessionDescriptor(uploadSession.SessionId);
```
Comment (xakep139) [inline src/VStore/Sessions/SessionManagementService.cs]: Further creation/modification of objects in such scenario will return status code 422 to client with violation of `binaryExists` template constraint

File: src/VStore/Sessions/SessionManagementService.cs (hunk 2 above)
Comment (xakep139) [inline src/VStore/Sessions/SessionManagementService.cs]: There definitely will be a time period between uploading binary data and committing the whole object, GC could easily delete binaries in this period.

File: src/VStore/Sessions/SessionManagementService.cs (hunk 4 above)
Comment (xakep139) [inline src/VStore/Sessions/SessionManagementService.cs]: At the moment we retreive session descriptor only for first uploading part. If we really need to check expiration of session, we could do this without querying Ceph.

File: src/VStore/Sessions/SessionManagementService.cs (hunk 2 above)
Comment (denisivan0v) [inline src/VStore/Sessions/SessionManagementService.cs]: Good point. Added the verification of a session expiration in #50

File: src/VStore/Sessions/SessionManagementService.cs (hunk 4 above)
Comment (denisivan0v) [inline src/VStore/Sessions/SessionManagementService.cs]: Agreed. Let's do the expiration checks on every file part uploading

File: src/VStore/Sessions/SessionManagementService.cs [hunk 5]
```diff
@@ -138,35 +145,30 @@ public sealed class SessionManagementService
...
-                                           Key = key,
-                                           UploadId = uploadSession.UploadId,
-                                           InputStream = memory,
-                                           PartNumber = uploadSession.NextPartNumber
-                                       });
-                uploadSession.AddPart(response.ETag);
-            }
+            var sessionDescriptor = uploadSession.SessionDescriptor;
+            var elementDescriptor = uploadSession.ElementDescriptor;
+            EnsureFileContentIsValid(
```
Comment (denisivan0v) [inline src/VStore/Sessions/SessionManagementService.cs]: Let's also split header and file content checks. Header will be checked in `UploadFilePart` if it is the first part. Content will be checked at `CompleteMultipartUpload`

Comment (xakep139) [inline src/VStore/Sessions/SessionManagementService.cs]: Done in https://github.com/2gis/nuclear-vstore/pull/69/commits/7cf75951418d5e617a65f6eb89a977fa1db76348

File: src/VStore.Host/Filters/MultipartBodyLengthLimitAttribute.cs [hunk 6]
```diff
@@ -2,26 +2,31 @@
...
         public void OnAuthorization(AuthorizationFilterContext context)
         {
+            if (_formOptions == null)
+            {
+                var options = context.HttpContext.RequestServices.GetService<IOptions<VStoreOptions>>();
+                _formOptions = new FormOptions
+                    {
+                        MultipartBodyLengthLimit = options.Value.MaxBinarySize
+                    };
+            }
```
Comment (denisivan0v) [inline src/VStore.Host/Filters/MultipartBodyLengthLimitAttribute.cs]: That is not thread safe. But I think we don't need thread safety here

Comment (xakep139) [inline src/VStore.Host/Filters/MultipartBodyLengthLimitAttribute.cs]: Agreed

**Comment by xakep139** [comment]
> Agreed. Let's do the expiration checks on every file part uploading

Need to certainly do this

**Comment by xakep139** [comment]
@denisivan0v, done session expiration checks in https://github.com/2gis/nuclear-vstore/pull/69/commits/fa26d3e47e9524bb13f651e4df0b84c11d3c38ea

File: src/VStore/Sessions/SessionStorageReader.cs [hunk 7]
```diff
@@ -111,6 +111,6 @@ public async Task<(SessionDescriptor SessionDescriptor, AuthorInfo AuthorInfo, D
             }
         }
 
-        private static bool IsSessionExpired(DateTime expiresAt) => expiresAt <= SessionDescriptor.CurrentTime();
+        public static bool IsSessionExpired(DateTime expiresAt) => expiresAt <= SessionDescriptor.CurrentTime();
```
Comment (denisivan0v) [inline src/VStore/Sessions/SessionStorageReader.cs]: Move it to the `SessionDescriptor` as static method?

Comment (xakep139) [inline src/VStore/Sessions/SessionStorageReader.cs]: Done in https://github.com/2gis/nuclear-vstore/pull/69/commits/02d12551525ec85612ae5a2b3a05b33b16e864db

**Pull request merged**